### ISSUE TITLE
How do I add a password when using 'uid'

### ISSUE THREAD

**Issue by zwhitchcox** [joyent/node#14188]
I am currently executing a child process, not as root, and trying to execute it as a different user.

Here is my sample code.

```js
var exec = require('child_process').exec,
    child;

child = exec('whoami',{uid:504},
  function (error, stdout, stderr) {
    console.log('stdout: ' + stdout);
    console.log('stderr: ' + stderr);
    if (error !== null) {
      console.log('exec error: ' + error);
    }
});
```

This doesn't work, however; I get this error:
```sh
child_process.js:1155
    throw errnoException(err, 'spawn');
          ^
Error: spawn EPERM
    at exports._errnoException (util.js:746:11)
    at ChildProcess.spawn (child_process.js:1155:11)
    at exports.spawn (child_process.js:988:9)
    at Object.exports.execFile (child_process.js:682:15)
    at exports.exec (child_process.js:642:18)
    at Object.<anonymous> (/Users/zane/test.js:4:9)
    at Module._compile (module.js:460:26)
    at Object.Module._extensions..js (module.js:478:10)
    at Module.load (module.js:355:32)
    at Function.Module._load (module.js:310:12)
```

I presume this is because I need to enter a password, but I don't really have a place to enter the password for this user. Is there any way to do this with node?

**Comment by zwhitchcox** [comment]
Solved it!

After five hours :'D

my code:

```js
  exec = require('child_process').exec
  child = exec('echo password | sudo -u world python packages/custom/algos/server/playground/test.py', (error, stdout, stderr) ->

    console.log('stdout: ' + stdout);
    console.log('stderr: ' + stderr);
    if error?
      console.log 'exec error: ' + error
  )
  child.stdout.on('data',  (data) ->
    console.log 'data: ' +data

  )
```

**Comment by misterdjules** [comment]
The `setuid` system call used by `child_process.exec` when the `uid` option is passed does not in itself interacts with users. On UNIX systems, most of the time PAM handles the interaction with users so that they can authenticate by, among other possibilities, providing a password.

Also note that programs such as `su` have the setuid bit set:
```
➜  v0.12 git:(v0.12) ✗ ls -l `which su`
-rwsr-xr-x  1 root  wheel  21488 Oct 30  2013 /usr/bin/su
➜  v0.12 git:(v0.12) ✗ 
```
This bit allows `su` to run with the elevated privileges necessary to be able to change the current user to any other user on the system.

In other words, if your node program needs to be able to change its user ID to any user, you will need to run with elevated privileges first. This is not recommended though. To provide users an interactive way to authenticate using the underlying system's facilities on UNIX, like by providing a password, one way to do that is to use a PAM module.

Closing this issue as this is not a problem with Node.js' core. If you need more help with doing that, please post your question in [the Node.js Google group](https://groups.google.com/forum/#!forum/nodejs) or on StackOverflow.

Thank you!
